### ISSUE TITLE
Generate a yaml file listing all techniques

### ISSUE THREAD

**Issue by mario-areias** [DataDog/stratus-red-team#172]
TL;DR This feature request is about to automatically generate a yaml file with listing all the techniques.  Similar to how atomic-red-team generates their [index file](https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/Indexes/index.yaml).

This project already has a list of techniques and documentation about the techniques. However, most of the documentation (to my knowledge) are stored in markdown files such as this one: https://github.com/DataDog/stratus-red-team/blob/main/docs/attack-techniques/AWS/aws.credential-access.ec2-get-password-data.md

I would like to propose have something similar but using yaml files. It would make easier for computers to consume yaml files than markdown files. Atomic-red-team has something similar in their project like this file [index.yaml](https://github.com/redcanaryco/atomic-red-team/blob/master/atomics/Indexes/index.yaml).

If this is something useful for stratus-red-team I can come up with a proposal and if it is accepted, I can create a PR implementing it.

**Comment by mario-areias** [comment]
## High level

Add the package https://pkg.go.dev/gopkg.in/yaml.v3 to stratus and annotate the `AttackTechnique` struct with the `yaml` annotation. Add the yaml serialisation on the `generate-techniques-documentation.go` file so the yaml index file would be generated at the same as any other documentation. The index would be saved on `docs/index.yaml`.   

##  Detailed proposal

Create a yaml index file with all techniques organised by platform, mitre attacks and attack techniques. That would be accomplished by serialiasing the `index` variable defined on `generate-techniques-documentation.go`. 

```go
	// Platform => [MITRE ATT&CK tactic => list of stratus techniques]
	index := map[stratus.Platform]map[string][]*stratus.AttackTechnique{}
```

https://pkg.go.dev/gopkg.in/yaml.v3 allows struct to be annotated make them easier to be serialised. `stratus.Platform` is a string field so it probably doesn't need to be annotated. The `stratus.AttackTechnique` struct needs to be annotated and there is a proposal below. Note that `yaml:"-"` means that field won't be serialised (or deserialised). 

```go
type AttackTechnique struct {
	// Short identifier, e.g. aws.persistence.create-iam-user
	ID string `yaml:"id"`

	// Friendly-looking short name
	FriendlyName string `yaml:"name"`

	// Full description (multi-line)
	Description string `yaml:"description"`

	// Pointer and leads for detection opportunities (multi-line)
	Detection string `yaml:"detection"`

	// Indicates if the technique is expected to be slow to warm-up or detonate
	IsSlow bool `yaml:"slow"`

	// MITRE ATT&CK Tactics to which this technique maps
	// see https://attack.mitre.org/techniques/enterprise/
	MitreAttackTactics []mitreattack.Tactic `yaml:"tactic"`

	// The platform of the technique, e.g. AWS
	Platform Platform `yaml:"platform"`

	// Terraform code to apply to create the necessary prerequisites for the technique to be detonated
	PrerequisitesTerraformCode []byte `yaml:"-"`

	// Detonation function
	// Parameters are the Terraform outputs
	Detonate func(params map[string]string) error `yaml:"-"`

	// Indicates if the detonation function is idempotent, i.e. if it can be run multiple times without reverting it
	IsIdempotent bool `yaml:"idempotent"`

	// Reversion function, to revert the side effects of a detonation
	Revert func(params map[string]string) error `yaml:"-"`
}
```

**Comment by christophetd** [comment]
Hello!

Sounds like a good idea, feel free to send a pull request! I'd happily accept a contribution. We can have the generation script in `v2/tools/`.

Small comments:
* I would rename `tactic` to `mitre-attack-tactic` to make it clear the terminology is coming from MITRE ATT&CK.
* Do we need the full description and detection? If that's only for meta-data, I don't think so. Let me know your thoughts

Thanks a lot!

**Comment by mario-areias** [comment]
@christophetd , I just created #218 - I tried to assign you as a reviewer but I don't think I have permissions to.
